### PR TITLE
Fluid channels should match the semantics of Go Channels

### PR DESCRIPTION
**Pull request by abhinavarora**
Fixes https://github.com/PaddlePaddle/Paddle/issues/8813

### CODE REVIEW

File: paddle/fluid/framework/channel_impl.h [hunk 1]
```diff
@@ -143,20 +142,24 @@ bool ChannelImpl<T>::Send(T *item) {
...
+    else {
       // We cannot do the data transfer because
       // this QueueMessage was added by Select
       // and some other case was executed.
       // So call the Send function again.
       // We do not care about notifying other
       // because they would have been notified
       // by the executed select case.
-      return send_return(Send(item));
+      Send(item);
```
Comment (cs2be) [inline paddle/fluid/framework/channel_impl.h]: Do you need to "lock.unlock();" here?

File: paddle/fluid/framework/channel_impl.h [hunk 2]
```diff
@@ -167,16 +170,20 @@ bool ChannelImpl<T>::Send(T *item) {
...
 
   // Block on channel, because some receiver will complete
   // the operation for us
   auto m = std::make_shared<QueueMessage>(item);
   sendq.push_back(m);
   m->Wait(lock);
-  // TODO(abhinavarora) Should panic on closed channel
-  return send_return(!m->chan_closed);
+  if (m->chan_closed) {
+    send_return();
```
Comment (cs2be) [inline paddle/fluid/framework/channel_impl.h]: Should unlock before throwing exception

File: paddle/fluid/framework/channel_impl.h [hunk 3]
```diff
@@ -167,16 +170,20 @@ bool ChannelImpl<T>::Send(T *item) {
...
+    PADDLE_THROW("Cannot send on closed channel");
+  }
+  send_return();
```
Comment (cs2be) [inline paddle/fluid/framework/channel_impl.h]: Do you need to unlock here?

File: paddle/fluid/framework/channel_impl.h [hunk 4]
```diff
@@ -118,15 +117,15 @@ bool ChannelImpl<T>::CanReceive() {
 }
 
 template <typename T>
-bool ChannelImpl<T>::Send(T *item) {
+void ChannelImpl<T>::Send(T *item) {
   send_ctr++;
   std::unique_lock<std::recursive_mutex> lock{mu_};
```
Comment (cs2be) [inline paddle/fluid/framework/channel_impl.h]: Do you need to explicitly lock after constructor?  lock->lock() ?

File: paddle/fluid/framework/channel_impl.h (hunk 1 above)
Comment (abhinavarora) [reply paddle/fluid/framework/channel_impl.h]: I think you are right, it makes sense to release the lock there. With the new semantics if the nested method call leads to an exception then the outer lock will be held forever.

File: paddle/fluid/framework/channel_impl.h (hunk 3 above)
Comment (abhinavarora) [reply paddle/fluid/framework/channel_impl.h]: Good catch, the lock needs to be unlocked here. Thank you for pointing this out.

File: paddle/fluid/framework/channel_impl.h (hunk 4 above)
Comment (abhinavarora) [reply paddle/fluid/framework/channel_impl.h]: No we don't need to do that. The unique lock constructor automatically does that.

File: paddle/fluid/framework/channel.h [hunk 5]
```diff
@@ -84,90 +84,89 @@ class ChannelHolder {
   }
 
   template <typename T>
-  bool Send(T* data) {
-    if (!IsInitialized()) return false;
+  void Send(T* data) {
+    PADDLE_ENFORCE_EQ(IsInitialized(), true);
```
[the remark this answers is not in the document]
Comment (abhinavarora) [reply paddle/fluid/framework/channel.h]: Done

**Pull request merged**